dali/kernels/common/split_shape.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_SPLIT_SHAPE_H_
+#define DALI_KERNELS_COMMON_SPLIT_SHAPE_H_
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Utility to divide a bigger shape into smaller blocks, given a desired number of blocks
+ *        and a minimum practical block size.
+ *        The algorithm start splitting from the outtermost dimension until the number of blocks
+ *        reaches the desired number or until the remaining volume is under a given threshold.
+ * @param split_factor Output argument used to represent split factors for each dimension.
+ * @param in_shape Input shape
+ * @param min_nblocks Desired minimum number of blocks
+ * @param min_sz Minimum practical block size
+ */
+template <typename SplitFactor, typename Shape>
+void split_shape(SplitFactor& split_factor, const Shape& in_shape, int min_nblocks = 8,
+                 int min_sz = (16 << 10)) {
```

Comment:
If this is intended for threading, then there are several issues:
1. If we can divide the shape into a number of threads that's a multiple of thread count, then we need no more partitions.
2. We should avoid uneven splits unless there's a good chance that it will be masked by the large number of blocks (much more than threads).

Reply:
My idea is that the caller will specify a suitable value for min_nblocks (a multiple of the number of threads)